spirv_cross/src/emscripten.rs

```diff
@@ -0,0 +1,142 @@
+    const MODULE_NAME: &'static str = "sc_internal";
+    let module = Reflect::get(&global(), &JsValue::from_str(MODULE_NAME)).unwrap().into();
+    Module {
+        module
+    }
+}
+
+const U32_SIZE: u32 = std::mem::size_of::<u32>() as u32;
+
+fn get_value<T>(object: &Object, key: &str) -> T 
+where T: std::convert::From<wasm_bindgen::JsValue> {
+    Reflect::get(object, &JsValue::from_str(key)).unwrap().into()
+}
+
+/// An Emscripten pointer.
+/// Internally stores an offset to a location on the Emscripten `u8` heap.
+#[derive(Clone, Copy)]
+pub struct Pointer {
+    offset: u32,
+}
+
+impl Pointer {
+    pub fn from_offset(offset: u32) -> Self {
+        Pointer {
+            offset,
+        }
+    }
+
+    pub fn as_offset(&self) -> u32 {
+        self.offset
+    }
+}
+
+pub struct Module {
+    module: Object,
+}
+
+impl Module {
+    /// Allocate memory on the heap.
+    pub unsafe fn allocate(&self, byte_len: u32) -> Pointer {
+        Pointer {
+            offset: _malloc(byte_len),
+        }
+    }
+
+    pub unsafe fn free(&self, pointer: Pointer) {
+        _free(pointer.as_offset())
+    }
+
+    // Read a `u32` value from the heap.
+    pub unsafe fn get_u32(&self, pointer: Pointer) -> u32 {
+        let offset = &JsValue::from_f64((pointer.offset / U32_SIZE) as f64);
+        // TODO: Remove Reflect
+        Reflect::get(&self.heap_u32(), offset).unwrap().as_f64().unwrap() as u32
+    }
+
+    /// Set memory on the heap to `bytes`.
+    pub unsafe fn set_from_u8_typed_array(&self, pointer: Pointer, bytes: Uint8Array) {
+        let buffer: JsValue = self.heap_u8().buffer().into();
+        let memory = Uint8Array::new_with_byte_offset_and_length(&buffer, pointer.offset, bytes.length());
+        memory.set(&bytes, 0);
+    }
+
+    /// Set memory on the heap to `bytes`.
+    pub unsafe fn set_from_u8_slice(&self, pointer: Pointer, bytes: &[u8]) {
+        self.set_from_u8_typed_array(pointer, Uint8Array::view(bytes));
+    }
+
+    fn heap_u8(&self) -> Uint8Array {
+        const HEAP_U8: &'static str = "HEAPU8";
+        get_value(&self.module, HEAP_U8)
+    }
+
+    fn heap_u32(&self) -> Uint32Array {
+        const HEAP_U32: &'static str = "HEAPU32";
+        get_value(&self.module, HEAP_U32)
+    }
+
+    /// Clones all bytes from the heap into a `Vec<u8>` while `should_continue` returns `true`.
+    /// Optionally include the last byte (i.e. to support peeking the final byte for nul-terminated strings).
+    pub unsafe fn read_bytes_into_vec_while<F>(&self, pointer: Pointer, should_continue: F, include_last_byte: bool) -> Vec<u8>
+    where F: Fn(u8, usize) -> bool {
+        let mut bytes = Vec::new();
+        let heap = &self.heap_u8();
+        let start_offset = pointer.offset as usize;
+        loop {
+            let bytes_read = bytes.len();
+            let offset = &JsValue::from_f64((start_offset + bytes_read) as f64);
+            let byte = Reflect::get(heap, offset).unwrap().as_f64().unwrap() as u8;
```

Comment:
The `Reflect::get` here returns a [`JsValue`](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen/struct.JsValue.html), and in order to read the value from JS we have to move through a `f64` (https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen/struct.JsValue.html#method.as_f64).

In general the real issue here is that we have to go through `f64` moving through the JS layer, because JS only knows `Number`. So even though we're trying to read from a `Uint8Array`, I don't see a way around this for now.

Although there is a slight improvement here for indexing. I asked about a better way to index into a typed array a while ago, and it appears the improvement has been merged since I originally wrote this: https://github.com/rustwasm/wasm-bindgen/pull/1147

Reply:
Now comes the dumb question! If we target WASM, why do we have to go through `JsValue`?

Reply:
This is the part about dynamic linking mentioned briefly in my comment. I'm not sure how to dynamically two link wasm modules and make them use the same memory space. So without having a way to dynamically link, we have to copy things back and forth between wasm modules, and move everything through JS to accomplish it instead.

I searched around for resources about this, but I couldn't find much except some C++ conventions to achieve dynamic linking between two C++ modules with some help from Emscripten. But it would be amazing if we could merge the wasm modules together somehow to avoid the JS layer: it would mean we can drop the majority of the code in this PR and have one code path for both native/wasm.

I don't know if host bindings helps will help with this and unfortunately I couldn't find much about the plans for this, except some wasm design documents describing how it might eventually be accomplished.

Reply:
Here's a reference document so we can follow the future progress of wasm dynamic linking: https://github.com/WebAssembly/tool-conventions/blob/master/DynamicLinking.md